<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>Python exception retry solution - SoByte</title>
  <meta name="renderer" content="webkit" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>

<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />

<meta name="theme-color" content="#f8f5ec" />
<meta name="msapplication-navbutton-color" content="#f8f5ec">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="#f8f5ec">


<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-6356451834813761" crossorigin="anonymous"></script>


<script async src="https://www.googletagmanager.com/gtag/js?id=G-E8GRRGBTEZ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-E8GRRGBTEZ');
</script>


<meta name="author" content="" /><meta name="description" content="When data crawling, we often encounter program exceptions due to network problems, and the initial approach is just to record the error content and post-process the error content again. Some better exception retry methods or mechanisms are compiled here.
Initial version :
1 2 3 4 5 6 7 8 9 10 11  def crawl_page(url): pass def log_error(url): pass url = &amp;#34;&amp;#34; try: crawl_page(url) except: log_error(url)   Improved version (increased number of retries):" /><meta name="keywords" content="Python, Error Retry" />






<meta name="generator" content="Hugo 0.92.2 with theme even" />


<link rel="canonical" href="https://www.sobyte.net/post/2021-11/python-error-retry/" />
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="manifest" href="/manifest.json">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">



<link href="/sass/main.min.f92fd13721ddf72129410fd8250e73152cc6f2438082b6c0208dc24ee7c13fc4.css" rel="stylesheet">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.css" integrity="sha256-7TyXnr2YU040zfSP+rEcz29ggW4j56/ujTPwjMzyqFY=" crossorigin="anonymous">


<meta property="og:title" content="Python exception retry solution" />
<meta property="og:description" content="When data crawling, we often encounter program exceptions due to network problems, and the initial approach is just to record the error content and post-process the error content again. Some better exception retry methods or mechanisms are compiled here.
Initial version :
1 2 3 4 5 6 7 8 9 10 11  def crawl_page(url): pass def log_error(url): pass url = &#34;&#34; try: crawl_page(url) except: log_error(url)   Improved version (increased number of retries):" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.sobyte.net/post/2021-11/python-error-retry/" /><meta property="article:section" content="post" />
<meta property="article:published_time" content="2021-11-16T22:43:53+08:00" />
<meta property="article:modified_time" content="2021-11-16T22:43:53+08:00" />

<meta itemprop="name" content="Python exception retry solution">
<meta itemprop="description" content="When data crawling, we often encounter program exceptions due to network problems, and the initial approach is just to record the error content and post-process the error content again. Some better exception retry methods or mechanisms are compiled here.
Initial version :
1 2 3 4 5 6 7 8 9 10 11  def crawl_page(url): pass def log_error(url): pass url = &#34;&#34; try: crawl_page(url) except: log_error(url)   Improved version (increased number of retries):"><meta itemprop="datePublished" content="2021-11-16T22:43:53+08:00" />
<meta itemprop="dateModified" content="2021-11-16T22:43:53+08:00" />
<meta itemprop="wordCount" content="571">
<meta itemprop="keywords" content="python," /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Python exception retry solution"/>
<meta name="twitter:description" content="When data crawling, we often encounter program exceptions due to network problems, and the initial approach is just to record the error content and post-process the error content again. Some better exception retry methods or mechanisms are compiled here.
Initial version :
1 2 3 4 5 6 7 8 9 10 11  def crawl_page(url): pass def log_error(url): pass url = &#34;&#34; try: crawl_page(url) except: log_error(url)   Improved version (increased number of retries):"/>

<!--[if lte IE 9]>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js"></script>
<![endif]-->

<!--[if lt IE 9]>
  <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
<![endif]-->

</head>
<body>
  <div id="mobile-navbar" class="mobile-navbar">
  <div class="mobile-header-logo">
    <a href="/" class="logo">SoByte</a>
  </div>
  <div class="mobile-navbar-icon">
    <span></span>
    <span></span>
    <span></span>
  </div>
</div>
<nav id="mobile-menu" class="mobile-menu slideout-menu">
  <ul class="mobile-menu-list">
    <a href="/">
        <li class="mobile-menu-item">Home</li>
      </a><a href="/post/">
        <li class="mobile-menu-item">Archives</li>
      </a><a href="/tags/">
        <li class="mobile-menu-item">Tags</li>
      </a><a href="/categories/">
        <li class="mobile-menu-item">Categories</li>
      </a><a href="/about/">
        <li class="mobile-menu-item">About</li>
      </a><a href="/ukraine/">
        <li class="mobile-menu-item">UKRAINE</li>
      </a>
  </ul>

  


</nav>

  <div class="container" id="mobile-panel">
    <header id="header" class="header">
        <div class="logo-wrapper">
  <a href="/" class="logo">SoByte</a>
</div>





<nav class="site-navbar">
  <ul id="menu" class="menu">
    <li class="menu-item">
        <a class="menu-item-link" href="/">Home</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/post/">Archives</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/tags/">Tags</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/categories/">Categories</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/about/">About</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/ukraine/">UKRAINE</a>
      </li>
  </ul>
</nav>

    </header>

    <main id="main" class="main">
      <div class="content-wrapper">
        <div id="content" class="content">
          <article class="post">
    
    <header class="post-header">
      <h1 class="post-title">Python exception retry solution</h1>

      <div class="post-meta">
        <span class="post-time"> 2021-11-16 22:43:53 </span>
        <div class="post-category">
            <a href="/categories/skills/"> skills </a>
            </div>
          <span class="more-meta"> 571 words </span>
          <span class="more-meta"> 3 mins read </span>
        
      </div>
    </header>

    <div class="post-toc" id="post-toc">
  <h2 class="post-toc-title">Contents</h2>
  <div class="post-toc-content">
    <nav id="TableOfContents"></nav>
  </div>
</div>
    <div class="post-content">
      <p>When data crawling, we often encounter program exceptions due to network problems, and the initial approach is just to record the error content and post-process the error content again. Some better exception retry methods or mechanisms are compiled here.</p>
<p><img src="https://cdn.jsdelivr.net/gh/b0xt/sobyte-images/2021/11/16/3568914e6bf549fda49321edfe7aadfe.png" alt=""></p>
<p><strong>Initial version</strong> :</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-py" data-lang="py"><span class="k">def</span> <span class="nf">crawl_page</span><span class="p">(</span><span class="n">url</span><span class="p">):</span>
    <span class="k">pass</span>

<span class="k">def</span> <span class="nf">log_error</span><span class="p">(</span><span class="n">url</span><span class="p">):</span>
    <span class="k">pass</span>

<span class="n">url</span> <span class="o">=</span> <span class="s2">&#34;&#34;</span>
<span class="k">try</span><span class="p">:</span>
   <span class="n">crawl_page</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
<span class="k">except</span><span class="p">:</span>
    <span class="n">log_error</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
</code></pre></td></tr></table>
</div>
</div><p><strong>Improved version (increased number of retries)</strong>:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-py" data-lang="py"><span class="n">attempts</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">success</span> <span class="o">=</span> <span class="kc">False</span>
<span class="k">while</span> <span class="n">attempts</span> <span class="o">&lt;</span> <span class="mi">3</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">success</span><span class="p">:</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">crawl_page</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
        <span class="n">success</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="n">attempts</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">attempts</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="k">break</span>
</code></pre></td></tr></table>
</div>
</div><p><strong>New solution: retrying</strong></p>
<p>retrying is a Python retry package that can be used to automatically retry segments that may fail to run. retrying provides a decorator function retry, and the decorated function will then be re-executed under a run failure condition, and by default will keep retrying as long as it keeps reporting errors.</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-py" data-lang="py"><span class="kn">import</span> <span class="nn">random</span>
<span class="kn">from</span> <span class="nn">retrying</span> <span class="kn">import</span> <span class="n">retry</span>

<span class="nd">@retry</span>
<span class="k">def</span> <span class="nf">do_something_unreliable</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&#34;Broken sauce, everything is hosed!!!111one&#34;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&#34;Awesome sauce!&#34;</span>

<span class="nb">print</span><span class="p">(</span><span class="n">do_something_unreliable</span><span class="p">())</span>
</code></pre></td></tr></table>
</div>
</div><p>If we run the have_a_try function, then it will not finish executing until random.randint returns 5, otherwise it will keep reexecuting.</p>
<p>retry can also accept some parameters, this can be seen from the initialization function of the Retrying class in the source code as optional parameters.</p>
<ul>
<li>stop_max_attempt_number: used to set the maximum number of attempts, after which the retry will be stopped</li>
<li>stop_max_delay: set to 10000, for example, then from the point in time when the decorated function starts to execute to the point in time when the function runs successfully or fails to abort, as long as this time exceeds 10 seconds, the function will not be executed again</li>
<li>wait_fixed: set the dwell time between retrying</li>
<li>wait_random_min and wait_random_max: generate the dwell time between retrying in a random way</li>
<li>wait_exponential_multiplier and wait_exponential_max: generates the dwell time between retrying in exponential form, generating a value of 2^previous_attempt_number * wait_exponential_ multiplier, previous_attempt_number is the number of previous retry, if the generated value exceeds the size of wait_exponential_max, then the dwell time between two retrying is wait_exponential_max. this design This design accommodates the exponential backoff algorithm and can mitigate blocking.</li>
<li>We can specify which exceptions we want to retry when they occur, this is done by passing a function object with retry_on_exception.</li>
</ul>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-py" data-lang="py"><span class="k">def</span> <span class="nf">retry_if_io_error</span><span class="p">(</span><span class="n">exception</span><span class="p">):</span>
    <span class="k">return</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">exception</span><span class="p">,</span> <span class="ne">IOError</span><span class="p">)</span>

<span class="nd">@retry</span><span class="p">(</span><span class="n">retry_on_exception</span><span class="o">=</span><span class="n">retry_if_io_error</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">read_a_file</span><span class="p">():</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&#34;file&#34;</span><span class="p">,</span> <span class="s2">&#34;r&#34;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
</code></pre></td></tr></table>
</div>
</div><p>During the execution of the read_a_file function, if an exception is reported, then this exception is passed into the retry_if_io_error function with the form parameter exception, if the exception is an IOError then retry is performed, if not it stops running and throws an exception.</p>
<p>We can also specify which results we want to retry when we get them, this is done by passing a function object with retry_on_result.</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-py" data-lang="py"><span class="k">def</span> <span class="nf">retry_if_result_none</span><span class="p">(</span><span class="n">result</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">result</span> <span class="ow">is</span> <span class="kc">None</span>

<span class="nd">@retry</span><span class="p">(</span><span class="n">retry_on_result</span><span class="o">=</span><span class="n">retry_if_result_none</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">get_result</span><span class="p">():</span>
    <span class="k">return</span> <span class="kc">None</span>
</code></pre></td></tr></table>
</div>
</div><p>After executing get_result successfully, the return value of the function will be passed into the retry_if_result_none function through the form parameter result, if the return value is None then retry will be performed, otherwise it will end and return the function value.</p>
<p>Reference.</p>
<ul>
<li><a href="https://pypi.python.org/pypi/retry/">https://pypi.python.org/pypi/retry/</a></li>
<li><a href="https://github.com/rholder/retrying">https://github.com/rholder/retrying</a></li>
<li><a href="https://julien.danjou.info/python-retrying/">Get back up and try again: retrying in Python</a></li>
</ul>

    </div>

    
<footer class="post-footer">
      <div class="post-tags">
          <a href="/tags/python/">python</a>
          </div>
      <nav class="post-nav">
        <a class="prev" href="/post/2021-11/go-benchmark-test/">
            <i class="iconfont icon-left"></i>
            <span class="prev-text nav-default">Write an interface pressure testing tool</span>
            <span class="prev-text nav-mobile">Prev</span>
          </a>
        <a class="next" href="/post/2021-11/requests-best-practice/">
            <span class="next-text nav-default">Tips for using the Python network request library Requests</span>
            <span class="next-text nav-mobile">Next</span>
            <i class="iconfont icon-right"></i>
          </a>
      </nav>
    </footer>
  </article>
        </div>
        

  

  

      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="social-links">
  <a href="https://www.sobyte.net/index.xml" type="application/rss+xml" class="iconfont icon-rss" title="rss"></a>
</div>

<div class="copyright">
  <span class="power-by">
    Powered by <a class="hexo-link" href="https://gohugo.io">Hugo</a>
  </span>
  <span class="division">|</span>
  <span class="theme-info">
    Theme - 
    <a class="theme-link" href="https://github.com/olOwOlo/hugo-theme-even">Even</a>
  </span>

  

  <span class="copyright-year">
    &copy; 
    2021 - 
    2022<span class="heart"><i class="iconfont icon-heart"></i></span><span></span>
  </span>
</div>

    </footer>

    <div class="back-to-top" id="back-to-top">
      <i class="iconfont icon-up"></i>
    </div>
  </div>
  
  <script src="https://cdn.jsdelivr.net/npm/jquery@3.2.1/dist/jquery.min.js" integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/slideout@1.0.1/dist/slideout.min.js" integrity="sha256-t+zJ/g8/KXIJMjSVQdnibt4dlaDxc9zXr/9oNPeWqdg=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.js" integrity="sha256-XVLffZaxoWfGUEbdzuLi7pwaUJv1cecsQJQqGLe7axY=" crossorigin="anonymous"></script>



<script type="text/javascript" src="/js/main.min.c99b103c33d1539acf3025e1913697534542c4a5aa5af0ccc20475ed2863603b.js"></script>
  <script type="text/javascript">
    window.MathJax = {
      tex: {
        inlineMath: [['$','$'], ['\\(','\\)']],
        tags: 'ams',
        }
    };
  </script>
  <script async src="https://cdn.jsdelivr.net/npm/mathjax@3.0.5/es5/tex-mml-chtml.js" integrity="sha256-HGLuEfFcsUJGhvB8cQ8nr0gai9EucOOaIxFw7qxmd+w=" crossorigin="anonymous"></script>








</body>
</html>
